writer.go

```diff
@@ -941,7 +941,7 @@ type partitionWriter struct {
 func newPartitionWriter(w *Writer, key topicPartition) *partitionWriter {
 	writer := &partitionWriter{
 		meta:  key,
-		queue: newBatchQueue(10),
+		queue: *newBatchQueue(10),
```

Comment:
Isn't this also a mutex copy?

Reply:
`go vet` doesn't complain about it, so perhaps not. If a pointer deference made a copy (in Go generally), then every dereference would produce a fresh copy and you'd never be able to work with the original. At least, that's my thinking, but I'm still new-ish to Go, so I'm happy to continue getting your insight!

Reply:
I don't believe the pointer dereference makes a copy but I think this initialization syntax turns into an assignment which would copy the dereferenced struct including the mutex.

Reply:
Oh, that's tricky! Thanks for the extra information!

Reply:
Found  https://cs.opensource.google/go/x/tools/+/refs/tags/v0.1.7:go/analysis/passes/copylock/testdata/src/a/copylock_func.go;l=119 
and 
https://github.com/golang/go/issues/16227

as well that talk about this exact case as well as copying zero valued locks being fine.

Reply:
@rhansen2 I reworked this change in a new commit back on Monday, so that this field is no longer initialized with a pointer dereference. But, I'm not sure now what kind of change you are in favor of, based on your last comment. Would you prefer that the code be reverted to copying mutexes as it originally did - since it's actually safe - even though vet complains about it?

Reply:
I'll defer to @achille-roussel  on this.

Reply:
I agree that the ideal code shape would be to no have a pointer-to-mutex, as this is an uncommon pattern (sometimes presented as an anti-pattern), so despite being correct it may be cause of concerns for future readers. 

An alternative approach could be to not have a constructor-style function, and instead use an initialization method:
```go
func (bq *batchQueue) init() {
  // do all the initialization here, no mutex copies happen
  ...
}
```

All that being said, a pragmatic move could be to accept the current change as it does fix the `go vet` warning and remains correct.

Up to you @bhavanki

Reply:
@achille-roussel I'll leave a comment with the concerns around introducing a pointer-to-mutex, and then merge what's here now. Someone can jump in later on and further revise the code to improve it more.